### PR TITLE
Measure and display latency in server view.

### PR DESCRIPTION
**Pull request by oskirby**
## Description
This adds storage of server latency data into the `Server` class, and adds a mechanism to refresh the server latency information whenever we update the server list, and then display that latency data to the user through a wifi with a variable number of bars.

## Reference
JIRA: [VPN-2526](https://mozilla-hub.atlassian.net/browse/VPN-2526)

## Checklist

- [x] My code follows the style guidelines for this project
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [ ] I have performed a self review of my own code
- [x] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

### CODE REVIEW

File: src/serverlatency.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+  connect(m_pingSender, SIGNAL(recvPing(quint16)), this,
+          SLOT(recvPing(quint16)));
+  connect(m_pingSender, SIGNAL(criticalPingError()), this,
+          SLOT(criticalPingError()));
+  connect(&m_timeout, &QTimer::timeout, this, &ServerLatency::stop);
+
+  // Send a ping to every server.
+  quint16 seq = 1;
+  for (const Server& server : scm->servers()) {
```
Comment (bakulf) [inline src/serverlatency.cpp:48]: I don't like the idea to ping all the servers at the same time. This means hundreds of ICMP packets and that can reach the rate limit on some OS.
I would suggest to use chunks of ICMP packets.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/serverlatency.cpp (hunk 1 above)
Comment (oskirby) [reply src/serverlatency.cpp:58]: Done!

File: src/models/servercountrymodel.h [hunk 2]
```diff
@@ -24,6 +24,13 @@ class ServerCountryModel final : public QAbstractListModel {
     CodeRole,
     CitiesRole,
   };
+  enum ServerConnectionScores {
+    Unavailable = -1,
+    NoData = 0,
+    Poor = 1,
+    Moderate = 2,
+    Good = 3,
+  };
```
Comment (bakulf) [inline src/models/servercountrymodel.h:33]: Q_ENUM ?

File: src/models/servercountrymodel.h [hunk 3]
```diff
@@ -74,10 +82,14 @@ class ServerCountryModel final : public QAbstractListModel {
...
   QVariant data(const QModelIndex& index, int role) const override;
 
+ signals:
+  void changed();
+
  private:
   [[nodiscard]] bool fromJsonInternal(const QByteArray& data);
 
   void sortCountries();
+  int cityConnectionScore(const ServerCity& city) const;
```
Comment (bakulf) [inline src/models/servercountrymodel.h:92]: ServerConnectionScore?

File: src/models/servercountrymodel.cpp [hunk 4]
```diff
@@ -156,22 +158,14 @@ QVariant ServerCountryModel::data(const QModelIndex& index, int role) const {
...
-        for (QString pubkey : city.servers()) {
-          if (m_servers.value(pubkey).cooldownTimeout() <= now) {
-            activeServerCount++;
-          }
-        }
-
         QStringList names{
             city.name(),
             ServerI18N::translateCityName(country.code(), city.name()),
-            QString::number(activeServerCount)};
```
Comment (bakulf) [inline src/models/servercountrymodel.cpp:174]: We should expose an object instead of this QStringList

**Review by bakulf — APPROVED** [review]

File: src/models/servercountrymodel.cpp (hunk 4 above)
Comment (oskirby) [reply src/models/servercountrymodel.cpp:174]: I agree, this feels quite messy and it makes it quite error-prone to extend the city data.

File: src/models/servercountrymodel.h (hunk 3 above)
Comment (oskirby) [reply src/models/servercountrymodel.h:92]: In this case we are actually scoring the city and not any one particular server at that location, so I feel that calling this a `cityConnectionScore` is slightly more accurate.

File: nebula/ui/components/VPNServerCountry.qml [hunk 5]
```diff
@@ -216,12 +216,13 @@ VPNClickableRow {
...
                     }
-                    source: "qrc:/nebula/resources/warning.svg"
+                    source: del.locationScore == 0 ? "qrc:/nebula/resources/warning.svg" :
+                        "qrc:/nebula/resources/wifi-" + (del.locationScore-1) + ".svg"
                     sourceSize {
                         height: VPNTheme.theme.iconSizeSmall
                         width: VPNTheme.theme.iconSizeSmall
                     }
-                    visible: !del.isAvailable
+                    visible: true
```
[the remark this answers is not in the document]
Comment (oskirby) [reply nebula/ui/components/VPNServerCountry.qml]: This prop is now used again when enabling and disabling the feature.

**Pull request merged**